tangelo/toolboxes/molecular_computation/rdms.py

```diff
@@ -292,3 +292,155 @@ def pad_rdms_with_frozen_orbitals(sec_mol, onerdm, twordm):
+    onerdm_b_padded[moidx_b[:, None], moidx_b] = onerdm_b
+
+    # Deleting the one rdm contribution in the two rdm. This must be done to
+    # redo the operation later with the one rdm with the frozen orbital.
+    twordm_aa = twordm_aa.transpose(1, 0, 3, 2)
+    twordm_bb = twordm_bb.transpose(1, 0, 3, 2)
+    twordm_ab = twordm_ab.transpose(1, 0, 3, 2)
+
+    onerdm_without_diag_a = np.copy(onerdm_a)
+    onerdm_without_diag_a[np.diag_indices(n_occ0_a)] -= 1
+    onerdm_without_diag_b = np.copy(onerdm_b)
+    onerdm_without_diag_b[np.diag_indices(n_occ0_b)] -= 1
+
+    onerdm_without_diag_a_T = onerdm_without_diag_a.T
+    onerdm_without_diag_b_T = onerdm_without_diag_b.T
+
+    for i in range(n_occ0_a):
+        twordm_aa[i, i, :, :] -= onerdm_without_diag_a
+        twordm_aa[:, :, i, i] -= onerdm_without_diag_a
+        twordm_aa[:, i, i, :] += onerdm_without_diag_a
+        twordm_aa[i, :, :, i] += onerdm_without_diag_a_T
+        twordm_ab[i, i, :, :] -= onerdm_without_diag_b
+
+    for i in range(n_occ0_b):
+        twordm_bb[i, i, :, :] -= onerdm_without_diag_b
+        twordm_bb[:, :, i, i] -= onerdm_without_diag_b
+        twordm_bb[:, i, i, :] += onerdm_without_diag_b
+        twordm_bb[i, :, :, i] += onerdm_without_diag_b_T
+        twordm_ab[:, :, i, i] -= onerdm_without_diag_a
+
+    for i, j in it.product(range(n_occ0_a), repeat=2):
+        twordm_aa[i, i, j, j] -= 1
+        twordm_aa[i, j, j, i] += 1
+
+    for i, j in it.product(range(n_occ0_b), repeat=2):
+        twordm_bb[i, i, j, j] -= 1
+        twordm_bb[i, j, j, i] += 1
+
+    for i in range(n_occ0_a):
+        for j in range(n_occ0_b):
+            twordm_ab[i, i, j, j] -= 1
+
+    # Creating a dummy two rdm.
+    dm2_aa = np.zeros((n_mos_a,)*4, dtype=twordm_aa.dtype)
+    dm2_bb = np.zeros((n_mos_b,)*4, dtype=twordm_bb.dtype)
+    dm2_ab = np.zeros((n_mos_a, n_mos_a, n_mos_b, n_mos_b), dtype=twordm_ab.dtype)
+
+    idx_a = (moidx_a.reshape(-1, 1) * n_mos_a + moidx_a).ravel()
+    idx_b = (moidx_b.reshape(-1, 1) * n_mos_b + moidx_b).ravel()
+
+    # This part is meant to replicate
+    # https://github.com/pyscf/pyscf/blob/8b3fef8cf18f10d430261d4a8bea21fadf19bb1f/pyscf/cc/uccsd_rdm.py#L561-L583.
+    # The output is not catched, maybe for memory efficiency purposes (elements
+    # of dm2 changed inplace?).
+    takebak_2d(dm2_aa.reshape(n_mos_a**2, n_mos_a**2),
+               twordm_aa.reshape(n_mos0_a**2, n_mos0_a**2), idx_a, idx_a)
+    takebak_2d(dm2_bb.reshape(n_mos_b**2, n_mos_b**2),
+               twordm_bb.reshape(n_mos0_b**2, n_mos0_b**2), idx_b, idx_b)
+    takebak_2d(dm2_ab.reshape(n_mos_a**2, n_mos_b**2),
+               twordm_ab.reshape(n_mos0_a**2, n_mos0_b**2), idx_a, idx_b)
+    twordm_aa_padded, twordm_ab_padded, twordm_bb_padded = dm2_aa, dm2_ab, dm2_bb
+
+    # The next few lines will reembed the on rdm, but with the frozen orbital
+    # elements of the one rdm matrix.
+    onerdm_a_padded_without_diag = np.copy(onerdm_a_padded)
```

Comment:
Think we can have a quick function defined in this function that avoids us repeating the same 30 lines of code ?
It looks like you did the same with the non-padded matrices already.

Reply:
I don't a way to be more concise at the moment. The goal is to "unpack" the RDM diagonals, reinstate missing columns and "repack" them at the end. This operation is done twice, but the first time in reverse (+ -> -, - -> +, etc.)

I don't think we would need those function elsewhere in a typical workflow. I also think it would impede readability to introduce (at least) 2 more functions to take into account the AA(BB) + AB terms in the 2-RDMs. In short, it requires some thoughts.

Reply:
Then it does not sound worth our time atm :)
Happy to leave it as that